src/java/main/org/apache/zookeeper/server/WatchManager.java

```diff
@@ -97,23 +102,44 @@ synchronized void removeWatcher(Watcher watcher) {
     Set<Watcher> triggerWatch(String path, EventType type, Set<Watcher> supress) {
         WatchedEvent e = new WatchedEvent(type,
                 KeeperState.SyncConnected, path);
-        HashSet<Watcher> watchers;
+        Set<Watcher> watchers = new HashSet<>();
+        PathIterator pathIterator = new PathIterator(path);
         synchronized (this) {
-            watchers = watchTable.remove(path);
-            if (watchers == null || watchers.isEmpty()) {
-                if (LOG.isTraceEnabled()) {
-                    ZooTrace.logTraceMessage(LOG,
-                            ZooTrace.EVENT_DELIVERY_TRACE_MASK,
-                            "No watchers for " + path);
+            for (String localPath : pathIterator.asIterable()) {
```

Comment:
It occurs to me that a check could be made to see if there are any Persistent watchers registered and skip if there aren't. I'll make that change.